### ISSUE TITLE
instaparse-live currently greedily walks all possible parse results

### ISSUE THREAD

**Issue by aengelberg** [mhuebert/instaparse-live#3]
This is a problem for parsers that return infinite parses.

Steps to reproduce:
- Clear the parse input box.
- Clear the grammar specification box, and begin typing: `S = '(' S ')' | S S | eps`.
- As you type the very last character (`s`), the browser will hang in an infinite loop.

This is because that seemingly simple parser for balanced parentheses returns infinite solutions for the empty string.

``` clojure
cljs.user=> (def p (insta/parser "S = '(' S ')' | S S | eps"))
S = "(" S ")" | S S | ε
cljs.user=> (p "")
[:S]
cljs.user=> (take 5 (insta/parses p ""))
([:S] [:S [:S] [:S]] [:S [:S] [:S [:S] [:S]]] [:S [:S [:S] [:S]] [:S [:S] [:S]]] [:S [:S [:S] [:S]] [:S]])
```

**Comment by mhuebert** [comment]
This happens because I use the insta/parses function to grab all possible parses - useful for removing ambiguity from a grammar. 

We could maintain this functionality with something like (take 100 ...), capping the max # of parses at 100 and removing the infinite loop. I tested this locally. 100 is probably more parses than you'd usually want to inspect visually/interactively?

**Comment by mhuebert** [comment]
- In the example you gave, 100 parses still stalls my computer noticeably. Do you think we could safely reduce this to something like 10 / 30 / 50?
  (Ultimately we may want to have a number of options but I haven't thought through a UI for this yet.)

**Comment by mhuebert** [comment]
Fixed in 910cf8565aa7418d2be04d36ad41591a3e9d8a6e by limiting parses to 20. Also improved formatting of results to make it more clear when multiple parses are being returned.

**Comment by aengelberg** [comment]
By adding the `(take 20 ...)`, you broke support for pretty-printing Failures on parse inputs (because `insta/failure?` doesn't work anymore, presumably). Now it just shows `No match` in red if the parse fails.

Speaking of which, is there any reason why the "no match" clause is in the code; I can't imagine why it would be called, because AFAIK the results aren't empty if and only if the parse isn't a failure.
